libraries/AP_Arming/AP_Arming.cpp

```diff
@@ -850,6 +850,116 @@ bool AP_Arming::camera_checks(bool display_failure)
+        aux_auth_fail_msg = (char *)calloc(aux_auth_str_len, sizeof(char));
+        if (aux_auth_fail_msg == nullptr) {
+            aux_auth_error = true;
+            return false;
+        }
+    }
+    auth_id = aux_auth_count;
+    aux_auth_state[auth_id] = AuxAuthStates::AUX_AUTH_STATE_NO_RESPONSE;
+    aux_auth_count++;
+    return true;
+}
+
+// set auxiliary authorisation passed
+void AP_Arming::set_aux_auth_passed(uint8_t auth_id)
+{
+    WITH_SEMAPHORE(aux_auth_sem);
+
+    // sanity check auth_id
+    if ((auth_id >= aux_auth_count_max) || (aux_auth_state[auth_id] == AuxAuthStates::AUX_AUTH_STATE_UNASSIGNED)) {
+        return;
+    }
+
+    aux_auth_state[auth_id] = AuxAuthStates::AUX_AUTH_STATE_PASSED;
+}
+
+// set auxiliary authorisation failed and provide failure message
+void AP_Arming::set_aux_auth_failed(uint8_t auth_id, const char* fail_msg)
+{
+    WITH_SEMAPHORE(aux_auth_sem);
+
+    // sanity check auth_id
+    if ((auth_id >= aux_auth_count_max) || (aux_auth_state[auth_id] == AuxAuthStates::AUX_AUTH_STATE_UNASSIGNED)) {
+        return;
+    }
+
+    // update state
+    aux_auth_state[auth_id] = AuxAuthStates::AUX_AUTH_STATE_FAILED;
+
+    // store failure message if this authoriser has the lowest auth_id
+    for (uint8_t i = 0; i < auth_id; i++) {
+        if (aux_auth_state[i] == AuxAuthStates::AUX_AUTH_STATE_FAILED) {
+            return;
+        }
+    }
+    if (aux_auth_fail_msg != nullptr) {
+        if (fail_msg == nullptr) {
+            strncpy(aux_auth_fail_msg, "Auxiliary authorisation refused", aux_auth_str_len);
+        } else {
+            strncpy(aux_auth_fail_msg, fail_msg, aux_auth_str_len);
+        }
+        aux_auth_fail_msg_source = auth_id;
+    }
+}
+
+bool AP_Arming::aux_auth_checks(bool display_failure)
+{
+    // handle error cases
+    if (aux_auth_error) {
+        if (aux_auth_fail_msg == nullptr) {
+            check_failed(ARMING_CHECK_AUX_AUTH, display_failure, "memory low for auxiliary authorisation");
+        } else {
+            check_failed(ARMING_CHECK_AUX_AUTH, display_failure, "Too many auxiliary authorisers");
+        }
+        return false;
+    }
+
+    WITH_SEMAPHORE(aux_auth_sem);
+
+    // check results for each auxiliary authorisation id
+    bool some_failed = false;
+    bool waiting_for_responses = false;
+    for (uint8_t i = 0; i < aux_auth_count; i++) {
+        switch (aux_auth_state[i]) {
+        case AuxAuthStates::AUX_AUTH_STATE_UNASSIGNED:
+            // do nothing
+            break;
+        case AuxAuthStates::AUX_AUTH_STATE_NO_RESPONSE:
+            waiting_for_responses = true;
+            break;
+        case AuxAuthStates::AUX_AUTH_STATE_FAILED:
+            if (i == aux_auth_fail_msg_source) {
+                check_failed(ARMING_CHECK_AUX_AUTH, display_failure, "%s", aux_auth_fail_msg);
+            }
+            some_failed = true;
+            break;
+        case AuxAuthStates::AUX_AUTH_STATE_PASSED:
+            break;
+        }
+    }
```

Comment:
You need to do a `check_failed` if there is an ID which is failing, but we are missing the string for it. At the moment its possible the check we had that set the failure string could have been set to passing, but another check is still failing, but hasn't updated itself recently, which means the failure string is stale.

Reply:
So the way it's written now, it will simply not display the message if it's stale but this is probably not good behaviour because it's been told to "display_failure" so I'll make it send a generic failure message like, "Waiting for auxiliary authorisation".

I'm not keen on fixing it further because, as discussed on the dev call last week, I suspect we will end up going back to adding authoriser independent failure messages which washes away all these issues.